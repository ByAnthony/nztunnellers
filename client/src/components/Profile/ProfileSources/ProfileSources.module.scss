@use 'src/index.scss';

.sources {
  margin-top: index.$base-margin;

  p {
    @include index.notes-paragraph
  }

  a {
    @include index.text-link;
  }
}
