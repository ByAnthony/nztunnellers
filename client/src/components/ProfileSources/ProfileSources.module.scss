@use 'src/index.scss';

.sources {
  margin: 2 * index.$base-margin 0;
}

.sources p {
  margin: 25px 0;
}

.sources ul {
  list-style: none;
}

.sources a {
  @include index.text-link;
}
