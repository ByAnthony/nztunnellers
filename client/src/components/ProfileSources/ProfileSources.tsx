import type {
  LondonGazette, NominalRoll, NzArchives, Sources,
} from '../../types/tunneller';
import STYLES from './ProfileSources.module.scss';

type Props = {
  sources: Sources;
}

type RecordWithIbid<T> = T & { ibid: string };

function addIbid<T extends Record<string, string>>(
  array: T[],
  index: number,
  ibid: string,
): RecordWithIbid<T>[] {
  return array.slice(index).map((obj) => ({ ...obj, ibid }));
}

function AwmmSource({ awmmCenotaph }: {awmmCenotaph: string}) {
  return (
    <p>
      {'Auckland War Memorial Museum Tāmaki Paenga Hira: '}
      <a href={awmmCenotaph}>Online Cenotaph He Toa Taumata Rau</a>
      .
    </p>
  );
}

function NzArchivesSource({ nzArchives }: {nzArchives: NzArchives[]}) {
  const nzArchivesList = [...addIbid([nzArchives[0]], 0, 'New Zealand Archives Te Rua Mahara o te Kāwanatanga, '), ...addIbid(nzArchives, 1, 'Ibid., ')];
  const italicIbid = (ibid: string) => (ibid === 'Ibid., ' ? <em>{ibid}</em> : ibid);
  return (
    <>
      {nzArchivesList.map((archives) => (
        <p key={archives.reference}>
          {italicIbid(archives.ibid)}
          {`${archives.reference}, `}
          <a href={archives.url}>Military Personnel File</a>
          .
        </p>
      ))}
    </>
  );
}

function LondonGazetteSource({ londonGazette }: {londonGazette: LondonGazette[]}) {
  if (londonGazette.length !== 0) {
    const LondonGazetteList = [...addIbid([londonGazette[0]], 0, 'London Gazette, '), ...addIbid(londonGazette, 1, 'Ibid., ')];
    return (
      <>
        {LondonGazetteList.map((gazette) => (
          <p key={gazette.page}>
            <em>{gazette.ibid}</em>
            {`${gazette.date}, p. ${gazette.page}`}
            .
          </p>
        ))}
      </>
    );
  }
  return null;
}

function NominalRollSource({ nominalRoll }: {nominalRoll: NominalRoll}) {
  const title = `${nominalRoll.title}`;
  const volumeRoll = `, ${nominalRoll.volume}, ${nominalRoll.roll}`;
  const reference = `, ${nominalRoll.publisher}, ${nominalRoll.town}, ${nominalRoll.date}, ${nominalRoll.page}.`;
  if (nominalRoll.volume && nominalRoll.roll) {
    return (
      <p>
        <em>{title}</em>
        {`${volumeRoll}${reference}`}
      </p>
    );
  }
  return (
    <p>
      <em>{title}</em>
      {reference}
    </p>
  );
}

export function ProfileSources({ sources }: Props) {
  return (
    <div className={STYLES.sources}>
      <h3>Sources</h3>
      <AwmmSource awmmCenotaph={sources.awmmCenotaph} />
      <NzArchivesSource nzArchives={sources.nzArchives} />
      <LondonGazetteSource londonGazette={sources.londonGazette} />
      <NominalRollSource nominalRoll={sources.nominalRoll} />
    </div>
  );
}
