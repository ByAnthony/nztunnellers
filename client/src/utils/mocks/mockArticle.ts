import { AboutUs } from '../../types/about-us';
import {
  Article, Image, Next, Section,
} from '../../types/article';

export const mockTitle: string = 'My Awesome\\Article Title';

export const mockChapter: number = 1;

export const mockTopImage: Image = {
  file: 'img-123.png',
  alt: 'Accessible alt text',
};

export const mockArticleImage: Image = {
  ...mockTopImage,
  title: 'An Amazing Photo',
  photographer: 'The French Photographer Doisneau',
  reference: 'Where the photo is preserved',
};

export const mockSectionOne: Section = {
  title: 'Section Title',
  text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit,sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea *commodo consequat*. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.\\n\\nLorem ipsum dolor sit amet, consectetur adipiscing elit,sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat[1](#footnote_1). Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.\\n\\n(6--July 2023)',
};

export const mockSectionTwo: Section = {
  title: 'Section Title',
  text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit,sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea *commodo consequat*. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.\\n\\nLorem ipsum dolor sit amet, consectetur adipiscing elit,sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat[1](#footnote_1). Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.\\n\\n(6--July 2023)',
};

export const mockNextButton: Next = {
  url: 'my-path-to-next-chapter/',
  title: 'Next\\Chapter',
  chapter: 3,
};

export const mockNotes: string = '[1.](#reference_1) Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.\\n\\n[2.](#reference_2) Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.';

export const mockArticle: Article = {
  id: 'my-awesome-article',
  chapter: mockChapter,
  title: mockTitle,
  section: [mockSectionOne, mockSectionTwo],
  image: [mockTopImage, mockArticleImage],
  next: mockNextButton,
  notes: mockNotes,
};

export const mockAboutUs: AboutUs = {
  id: 'my-awesome-article',
  title: mockTitle,
  section: [mockSectionOne, mockSectionTwo],
  image: [mockTopImage],
};
