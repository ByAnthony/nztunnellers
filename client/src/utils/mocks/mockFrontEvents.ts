import { Date, Event, EventDetail } from '../../types/tunneller';

const mockDate = (year: string | null, dayMonth: string | null): Date => ({
  year,
  dayMonth,
});

export const mockEvent: EventDetail = {
  description: null,
  title: null,
  image: null,
};

export const mockEventCompany: EventDetail = {
  ...mockEvent,
  description: 'Something happened that day for the Company',
  title: 'The Company',
  image: 'image.jpg',
};
export const mockEventDescription: EventDetail = { ...mockEvent, description: 'Something happened that day' };
export const mockEventEnlisted: EventDetail = { ...mockEvent, description: 'Main Body', title: 'Enlisted' };
export const mockEventPosted: EventDetail = { ...mockEvent, description: 'Reinforcement', title: 'Posted' };
export const mockEventTrained: EventDetail = { ...mockEvent, description: 'Military Camp, Somewhere', title: 'Trained' };
export const mockEventTitleAndDescription: EventDetail = { ...mockEvent, description: 'Something happened', title: 'Major Event' };

export const mockEventDiedOfDisease: EventDetail = { ...mockEvent, description: 'Military Hospital', title: 'Died of disease' };
export const mockEventKilledInAction: EventDetail = { ...mockEvent, description: 'Killed on the battlefield', title: 'Killed in action' };
export const mockEventDiedOfWounds: EventDetail = { ...mockEvent, description: 'Wounded on the battlefield', title: 'Died of wounds' };
export const mockEventDiedOfAccident: EventDetail = { ...mockEvent, description: 'Fatal accident', title: 'Died of accident' };
export const mockEventBuried: EventDetail = { ...mockEvent, description: 'Cemetery, France', title: 'Buried' };
export const mockEventGrave: EventDetail = { ...mockEvent, description: '1 A 26', title: 'Grave reference' };

const mockDeathEvent = (deathEvent: EventDetail): Record<string, Event[]> => ({
  1916: [
    {
      date: mockDate('1916', '27 September'),
      event: [mockEventCompany],
    },
    {
      date: mockDate('1916', '19 November'),
      event: [mockEventDescription],
    },
  ],
  1917: [
    {
      date: mockDate('1917', '8 August'),
      event: [mockEventDescription],
    },
    {
      date: mockDate('1917', '8 August'),
      event: [deathEvent, mockEventBuried, mockEventGrave],
    },
  ],
});

export const mockFrontEventsWithDiedOfDisease:
    Record<string, Event[]> = mockDeathEvent(mockEventDiedOfDisease);

export const mockFrontEventsWithKilledInAction:
    Record<string, Event[]> = mockDeathEvent(mockEventKilledInAction);

export const mockFrontEventsWithDiedOfWounds:
    Record<string, Event[]> = mockDeathEvent(mockEventDiedOfWounds);

export const mockFrontEventsWithDiedOfAccident:
    Record<string, Event[]> = mockDeathEvent(mockEventDiedOfAccident);

export const mockFrontEvents: Record<string, Event[]> = {
  1915: [
    {
      date: mockDate('1915', '26 January'),
      event: [mockEventEnlisted],
    },
    {
      date: mockDate('1915', '30 January'),
      event: [mockEventPosted],
    },

    {
      date: mockDate('1915', '14 February'),
      event: [mockEventTrained],
    },
    {
      date: mockDate('1915', '4 May'),
      event: [mockEventTitleAndDescription],
    },
  ],
  1916: [
    {
      date: mockDate('1916', '23 June'),
      event: [mockEventDescription, mockEventDescription],
    },
    {
      date: mockDate('1916', '27 September'),
      event: [mockEventCompany],
    },
    {
      date: mockDate('1916', '19 November'),
      event: [mockEventDescription],
    },
    {
      date: mockDate('1916', '12 December'),
      event: [mockEventCompany, mockEventDescription],
    },
  ],
  1917: [
    {
      date: mockDate('1917', '9 April'),
      event: [mockEventTitleAndDescription, mockEventCompany],
    },
    {
      date: mockDate('1917', '8 August'),
      event: [mockEventDescription],
    },
  ],
  1918: [
    {
      date: mockDate('1918', '3 March April'),
      event: [mockEventTitleAndDescription],
    },
    {
      date: mockDate('1918', '13 October'),
      event: [mockEventDescription],
    },
  ],
  1919: [
    {
      date: mockDate('1919', '4 April'),
      event: [mockEventTitleAndDescription],
    },
    {
      date: mockDate('1919', '12 July'),
      event: [mockEventDescription],
    },
  ],
};
