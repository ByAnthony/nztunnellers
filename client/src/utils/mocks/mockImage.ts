import {
  Image, ImageSource, ImageNewspaper, ImageArchives, ImageBook,
} from '../../types/tunneller';

export const mockImageAucklandLibraries: string = 'https://digitalnz.org/records?text=31-B2671&tab=Images#';

export const mockImageArchives: ImageArchives = {
  location: 'Auckland War Memorial Museum',
  reference: 'MS-93/157',
};

export const mockImageFamily: string = 'Courtesy of John Doe family';

export const mockImageNewspaper: ImageNewspaper = {
  name: 'Auckland Star',
  date: '12 July 1898',
};

export const mockImageBook: ImageBook = {
  title: 'My Mock Title',
  town: 'Bethune',
  publisher: 'Publisher and Co.',
  authors: [
    {
      forename: 'Jane',
      surname: 'Doe',
    },
  ],
  year: '1925',
  page: 'p. 89',
};

export const mockImageSource: ImageSource = {
  aucklandLibraries: null,
  archives: null,
  family: null,
  newspaper: null,
  book: null,
};

export const mockImage: Image = {
  url: '1/1234.jpg',
  source: mockImageSource,
};
