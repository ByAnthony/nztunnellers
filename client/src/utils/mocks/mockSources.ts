import {
  NzArchives, NominalRoll, LondonGazette, Sources,
} from '../../types/tunneller';

const mockNzArchives: NzArchives[] = [{
  reference: 'AABK 18805 W5530 39/0022386',
  url: 'https://www.mockurl.co.nz/nzarchives/B2874930',
}];

const mockAwmmCenotaph: string = 'https://www.mockurl.co.nz/online-cenotaph/B2874930';

const mockNominalRoll: NominalRoll = {
  title: 'Nominal Roll of New Zealand Expeditionary Force, 1915. New Zealand Engineers Tunnelling Company',
  town: 'Wellington',
  publisher: 'Government Printer',
  date: '1916',
  page: '37',
  volume: null,
  roll: null,
};

const mockLondonGazette: LondonGazette[] = [{
  page: '1675',
  date: '23 May 1917',
}];

export const mockSources = (overrides: Partial<Sources> | undefined = undefined): Sources => ({
  nzArchives: mockNzArchives,
  awmmCenotaph: mockAwmmCenotaph,
  nominalRoll: mockNominalRoll,
  londonGazette: mockLondonGazette,
  ...overrides,
});
