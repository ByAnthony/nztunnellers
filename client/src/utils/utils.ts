export const displayBirthDeathDates = (
  birth: string | null,
  death: string | null,
) => {
  if (birth !== null && death !== null) {
    return `${birth} - ${death}`;
  }
  if (birth !== null && death === null) {
    return `${birth} - †?`;
  }
  if (birth === null && death !== null) {
    return `? - ${death}`;
  }
  return '? - †?';
};
