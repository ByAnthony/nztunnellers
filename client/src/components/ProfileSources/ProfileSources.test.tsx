import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { mockSources } from '../../utils/mocks/mockSources';
import { ProfileSources } from './ProfileSources';
import { findElementWithText } from '../../utils/test-utils';

const component = <ProfileSources sources={mockSources()} />;

test('renders the component correctly', () => {
  const { asFragment } = render(component);

  expect(asFragment()).toMatchSnapshot();
});

test('renders Auckland War Memorial Museum information', () => {
  render(component);

  const element = findElementWithText('Auckland War Memorial Museum Tāmaki Paenga Hira');

  expect(element).toBeInTheDocument();
  expect(screen.getByRole('link', { name: 'Online Cenotaph He Toa Taumata Rau' }))
    .toHaveAttribute('href', 'https://www.mockurl.co.nz/online-cenotaph/B2874930');
});

test('renders NZ Archives information', () => {
  render(component);

  const element = findElementWithText('New Zealand Archives Te Rua Mahara o te Kāwanatanga, AABK 18805 W5530 39/0022386, ');

  expect(element).toBeInTheDocument();
  expect(screen.getByRole('link', { name: 'Military Personnel File' }))
    .toHaveAttribute('href', 'https://www.mockurl.co.nz/nzarchives/B2874930');
});

test('renders NZ Archives information list', () => {
  const componentWithExtraInformation = (
    <ProfileSources sources={mockSources({
      nzArchives: [
        {
          reference: 'AABK 18805 W5530 39/0022386',
          url: 'https://www.mockurl.co.nz/nzarchives/B2874930',
        },
        {
          reference: 'AABK 18805 W5530 39/002286',
          url: 'https://www.mockurl.co.nz/nzarchives/B284930',
        },
      ],
    })}
    />
  );

  render(componentWithExtraInformation);

  expect(screen.getByText('Ibid.,')).toBeTruthy();
});

test('renders London Gazette information', () => {
  render(component);

  const element = findElementWithText('23 May 1917, p. 1675.');

  expect(element).toBeInTheDocument();
});

test('renders London Gazette information list', () => {
  const componentWithExtraInformation = (
    <ProfileSources sources={mockSources({
      londonGazette: [
        {
          page: '1675',
          date: '23 May 1917',
        },
        {
          page: '457',
          date: '27 September 1919',
        },
      ],
    })}
    />
  );

  render(componentWithExtraInformation);

  expect(screen.getByText('Ibid.,')).toBeTruthy();
});

test('renders Nominal Roll information', () => {
  render(component);

  expect(screen.getByTestId('nominal-roll'))
    .toHaveTextContent('Nominal Roll of New Zealand Expeditionary Force, 1915. New Zealand Engineers Tunnelling Company, Government Printer, Wellington, 1916, 37.');
});

test('renders Nominal Roll with volume and roll information', () => {
  const componentWithExtraInformation = (
    <ProfileSources sources={mockSources({
      nominalRoll: {
        title: 'Nominal Rolls of New Zealand Expeditionary Force',
        town: 'Wellington',
        publisher: 'Government Printer',
        date: '1916',
        page: '37',
        volume: 'Volume III',
        roll: 'No.55',
      },
    })}
    />
  );

  render(componentWithExtraInformation);

  expect(screen.getByTestId('nominal-roll'))
    .toHaveTextContent('Nominal Rolls of New Zealand Expeditionary Force, Volume III, No.55, Government Printer, Wellington, 1916, 37.');
});
