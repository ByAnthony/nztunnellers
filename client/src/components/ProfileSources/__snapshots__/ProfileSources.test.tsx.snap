// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component correctly 1`] = `
<DocumentFragment>
  <div
    class="sources"
  >
    <h3>
      Sources
    </h3>
    <p
      data-testid="awmm"
    >
      Auckland War Memorial Museum Tāmaki Paenga Hira: 
      <a
        href="https://www.mockurl.co.nz/online-cenotaph/B2874930"
      >
        Online Cenotaph He Toa Taumata Rau
      </a>
      .
    </p>
    <p
      data-testid="nz-archives"
    >
      New Zealand Archives Te Rua Mahara o te Kāwanatanga, AABK 18805 W5530 39/0022386, 
      <a
        href="https://www.mockurl.co.nz/nzarchives/B2874930"
      >
        Military Personnel File
      </a>
      .
    </p>
    <p
      data-testid="london-gazette"
    >
      <em>
        London Gazette, 
      </em>
      23 May 1917, p. 1675.
    </p>
    <p
      data-testid="nominal-roll"
    >
      <em>
        Nominal Roll of New Zealand Expeditionary Force, 1915. New Zealand Engineers Tunnelling Company
      </em>
      , Government Printer, Wellington, 1916, 37.
    </p>
  </div>
</DocumentFragment>
`;
